OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -213,18 +222,33 @@ public void NotifyBuildingComplete(Actor self)
 
 		public void Selling(Actor self)
 		{
+			if (Info.RemoveSmudgesOnSell)
+				RemoveSmudges();
+
 			BuildComplete = false;
 		}
 
 		public void Sold(Actor self) { }
 
 		public void BeforeTransform(Actor self)
 		{
+			if (Info.RemoveSmudgesOnTransform)
+				RemoveSmudges();
+
 			foreach (var s in Info.UndeploySounds)
 				Game.Sound.PlayToPlayer(self.Owner, s, self.CenterPosition);
 		}
 
 		public void OnTransform(Actor self) { }
 		public void AfterTransform(Actor self) { }
+
+		public void RemoveSmudges()
+		{
+			var smudgeLayers = self.World.WorldActor.TraitsImplementing<SmudgeLayer>();
+
+			foreach (var smudgeLayer in smudgeLayers)
+				foreach (var occupiedCell in OccupiedCells())
```

Comment:
You probably want to use `FootprintUtils.Tiles` instead of `FootprintUtils.UnpathableTiles` (indirectly via `OccupiedCells`).